File: 127-word-ladder/word-ladder.cpp
class Solution {
public:

    // vector<string> adjacentWords(unordered_set<string>& wordSet, string curr)
    // {
    //     vector<string> ans;
    //     for(auto word : wordSet)
    //     {
    //         int diff=0;
    //         for(int i=0;i<curr.length();i++)
    //         {
    //             if(curr[i]!=word[i]) diff++;
    //         }
    //         if(diff==1) ans.push_back(word);
    //     }
    //     return ans;
    // }

    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        if(beginWord==endWord) return 0;
        unordered_set<string> wordSet;
        for(string word : wordList) wordSet.insert(word);
        if(wordSet.count(endWord)==0) return 0;
        queue<string> q;
        q.push(beginWord);
        if(wordSet.count(beginWord)!=0) wordSet.erase(beginWord);
        int step=1;
        while(!q.empty())
        {
            int stepSize = q.size();
            while(stepSize--)
            {
                string curr = q.front();
                q.pop();
                if(curr == endWord) return step;
                for(int i=0;i<curr.length();i++)
                {
                    char ogchar= curr[i];
                    for(int j=0;j<26;j++)
                    {
                        if(('a'+j)!=ogchar)
                        {
                            string newWord = curr;
                            newWord[i]='a'+j;
                            if(wordSet.count(newWord)!=0)
                            {
                                q.push(newWord);
                                wordSet.erase(newWord);
                            }
                        }
                    }
                }
                // for(auto word : adjacentWords(wordSet,curr))
                // {
                //     wordSet.erase(word);
                //     q.push(word);
                // } 
            }
            step++;
        }
        return 0;
    }
};

File: 127-word-ladder/word-ladder.java
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        Set<String> set = new HashSet<>();
        for(String word : wordList) set.add(word);
        if(!set.contains(endWord)) return 0;
        Deque<String> dq = new ArrayDeque<>();
        dq.addLast(beginWord);
        if(set.contains(beginWord)) set.remove(beginWord);
        int ans = 1;

        while(dq.size()>0)
        {
            int size = dq.size();
            while(size-- >0)
            {
                String currWord = dq.removeFirst();
                if(currWord.equals(endWord)) return ans;
                //generate next words
                for(int i=0;i<currWord.length();i++)
                {
                    for(int j=0;j<26;j++)
                    {
                        String genWord = currWord.substring(0,i) + (char)('a'+j) + currWord.substring(i+1,currWord.length());
                        if(!genWord.equals(currWord) && set.contains(genWord))
                        {
                            dq.addLast(genWord);
                            set.remove(genWord);
                        } 
                    }
                }
                
            }
            ans++;
        }

        return 0;
        
    }
}